From 3f588e6de9c64225ca7f3d2df4e2e46e72ab8dc9 Mon Sep 17 00:00:00 2001
From: Zhuocheng Ding <zhuocheng.ding@intel.com>
Date: Fri, 11 Sep 2020 09:32:03 +0800
Subject: [PATCH] mc146818rtc: add rtc-refresh-timer QMP command

If host is suspended for a while, the timers in QEMU will not take the
time during suspension into account. This command will force QEMU to
recheck RTC timers, so that these timers will not be delayed.

Tracked-On: OAM-92864
Signed-off-by: Zhuocheng Ding <zhuocheng.ding@intel.com>
---
 hw/rtc/mc146818rtc.c  |  6 ++++++
 qapi/misc-target.json | 14 ++++++++++++++
 2 files changed, 20 insertions(+)

diff --git a/hw/rtc/mc146818rtc.c b/hw/rtc/mc146818rtc.c
index 4db9e946d1..5ed0f7b100 100644
--- a/hw/rtc/mc146818rtc.c
+++ b/hw/rtc/mc146818rtc.c
@@ -45,6 +45,7 @@

 #ifdef TARGET_I386
 #include "qapi/qapi-commands-misc-target.h"
+#include "qemu/main-loop.h"
 #include "hw/i386/apic.h"
 #endif

@@ -119,6 +120,11 @@ void qmp_rtc_reset_reinjection(Error **errp)
     }
 }

+void qmp_rtc_refresh_timer(Error **errp)
+{
+    qemu_notify_event();
+}
+
 static bool rtc_policy_slew_deliver_irq(RTCState *s)
 {
     apic_reset_irq_delivered();
diff --git a/qapi/misc-target.json b/qapi/misc-target.json
index a92959d436..04bdd52eca 100644
--- a/qapi/misc-target.json
+++ b/qapi/misc-target.json
@@ -61,6 +61,20 @@
 { 'command': 'rtc-reset-reinjection',
   'if': 'defined(TARGET_I386)' }

+##
+# @rtc-refresh-timer:
+#
+# This command will refresh the RTC timers, so that the expired timers can
+# trigger their callbacks immediately.
+#
+# Example:
+#
+# -> { "execute": "rtc-refresh-timer" }
+# <- { "return": {} }
+#
+##
+{ 'command': 'rtc-refresh-timer',
+  'if': 'defined(TARGET_I386)' }

 ##
 # @SevState:
-- 
2.25.1

